src/aviary/functional.py

```diff
@@ -0,0 +1,166 @@
+        self.start_fn = start_fn
+        self.tools = tools
+        self.args = args
+        self.kwargs = kwargs
+        self.state = DynamicState()
+
+    async def _call_start(self) -> tuple[str, dict[str, Any]]:
+        """Call start_fn handling both sync and async cases."""
+        if is_coroutine_callable(self.start_fn):
+            result = await self.start_fn(*self.args, **self.kwargs)
+        else:
+            result = self.start_fn(*self.args, **self.kwargs)
+
+        if not isinstance(result, tuple) or len(result) != 2:  # noqa: PLR2004
+            raise TypeError("Start function must return (observation, state_dict)")
+
+        obs, state_dict = result
+        if not isinstance(state_dict, dict):
+            raise TypeError("State must be a dictionary")
+
+        return obs, state_dict
+
+    async def reset(self) -> tuple[Messages, list[Tool]]:
+        obs, extras = await self._call_start()
+        self.state = DynamicState.from_extras(extras)
+        return [Message(content=obs, role="user")], self.tools
+
+    async def step(
+        self, action: ToolRequestMessage
+    ) -> tuple[Messages, float, bool, bool]:
+        # just assume no concurrent to avoid user bugs
+        # TODO: could take config arguments in start() to enable concurrency
+        msgs = await self.exec_tool_calls(action, state=self.state, concurrency=False)
+        return msgs, self.state.reward, self.state.done, False
+
+    def export_frame(self) -> Frame:
+        """Export the current state of the environment."""
+        return Frame(
+            state={
+                "done": self.state.done,
+                "reward": self.state.reward,
+                "extras": self.state.extras,
+            },
+            info={
+                "tool_names": [t.info.name for t in self.tools],
+                "num_tools": len(self.tools),
+            },
+        )
+
+
+class EnvironmentBuilder:
+    """Builder class for constructing functional environments."""
+
+    def __init__(self, start_fn: Callable):
+        self.start_fn = start_fn
+        self.tools: list[Tool] = []
+        update_wrapper(self, start_fn)
+
+    def __call__(self, *args, **kwargs):
+        """Create a new environment instance."""
+        return FunctionalEnvironment(self.start_fn, self.tools.copy(), *args, **kwargs)
+
+    def tool(self, **tool_kwargs):
+        """Decorator to add tools to the environment."""
+
+        def decorator(func: Callable):
+            tool = Tool.from_function(
+                func, **tool_kwargs, allow_empty_param_descriptions=True
+            )
+            # Check for name conflicts
+            if any(t.info.name == tool.info.name for t in self.tools):
+                raise RuntimeError(f"Tool with name '{tool.info.name}' already exists")
+            self.tools.append(tool)
+            return func
+
+        return decorator
+
+
+class fenv:  # noqa: N801
+    """Factory class for creating functional environments."""
```

Comment:
Can you move this to be a `classmethod` of `EnvironmentBuilder`, and then at the module level set `fenv = EnvironmentBuilder.fenv`

Reply:
Feels aesthetic, will probably skip this.